src/option_parser.cr

```diff
@@ -368,19 +368,31 @@ class OptionParser
         if (handler = @handlers[flag]?) && !(handler.value_type.none? && value)
           handled_args << arg_index
 
-          # Pull in the next argument if we don't already have it and an argument
-          # is taken (i.e. not FlagValue::None)
-          if !value && !handler.value_type.none?
-            value = args[arg_index + 1]?
-            if value
-              handled_args << arg_index + 1
-              arg_index += 1
+          if !value
+            case handler.value_type
+            when FlagValue::Required
+              value = args[arg_index + 1]?
+              if value
+                handled_args << arg_index + 1
+                arg_index += 1
+              else
+                @missing_option.call(flag)
+              end
+            when FlagValue::Optional
+              value = args[arg_index + 1]?
+              # e.g. Do not match `-g` or `--g-long` but do match `--`
+              # NotThis|NotThat|GoAway|(WeWantThis)
+              if value && value =~ /^\-(?!\-)|^\-\-\S|^(.+)/ && $1?
```

Comment:
I don't follow the reasoning behind the regex. Doesn't the last clause (`^(.+)`) make it match *any* string (except the empty one)?

And what is `NotThis|NotThat|GoAway|(WeWantThis)` supposed to mean?

Reply:
`NotThis|NotThat|GoAway|(WeWantThis)` is a description of the regex. Every clause prior to the last one is something not to match on, the last clause matches things that are wanted. Thus, the two can be differentiated by whether they were captured or not. That's why the `$1` is the last part of the conditional.

A longer explanation is given here, it's called [The Greatest Regex Trick Ever](https://rexegg.com/regex-best-trick.html) (a title I agree with) and is designed to be the simplest way to do this kind of job. If there's a better way, I'm all ears.

Reply:
Ah I think I understand it. But I'm not particularly fond of using a regex that's really hard to reason about.

But before looking into possible alternative expressions, I'd like to talk about the intention.
Could it just be `!value.starts_with?("-")`? Why is `--` explicitly excluded?

Reply:
The complexity of the problem is increased because of handling the `--`. That's a special case because of #8937, specifically a case like this:

> But to do it correctly, the behavior must depend on context. If we're currently expecting an option value, the next token must be recognized as that regardless of its content.

> E.g. Git recognizes this as an option correctly:
> `git commit -m --` (create a commit with the message "--")

It's quite possible there's a simpler regex but I'd say that's straightforward to spot once you know that trick. It's more likely that it can be simplified by not using a regex and using more code.

How about using the whitespace regex flag to add in some comments? From line 382…

```suggestion
              value = args[arg_index + 1]?
              # We don't want short or long flags as values here
              # but sometimes double dashes alone are ok.
              # The easiest way to handle is to *capture*
              # what we want and drop all other matches.
              # e.g. not `-g` or `--g-long` but do capture `--`
              # The logic is:
              # `/NotThis|NotThat|AndNotThis...|(WeWantThis)/`
              flag_pattern = /
                ^\-(?!\-) # This matches short flags e.g. -g
                  |
                ^\-\-\S   # This matches long flags
                  |
                ^(.+)     # This captures what we want
              /x

              # If there is a capture ($1 is true)
              # then this condition is true
              if value && value =~ flag_pattern && $1?
```

The specs pass for that.

Reply:
Thanks, the explanatory comments are really helpful.

But I'm pretty sure a non-regex implementation would be easier to understand and even more efficient.
I believe `!value.starts_with?("-") || value == "-" || value == "--"` should be pretty much equivalent.

Reply:
Well, I'm not going to die on hill of "regex are easy to grok" 😆 though I do use them an awful lot, but your questions led me to a (hopefully) much better solution.

I realised that the regex would fail if something *looked* like an option e.g. `--not-an-option--`, and I added a spec for that and it did indeed fail. The answer…

```
              if value && !@handlers.has_key?(value)
```

Just check the handler's hash instead, all the flags are in there already! It passes all the specs..

Reply:
Yeah, I suppose that's probably the best solution.